localstack/services/apigateway/provider_asf.py

```diff
@@ -0,0 +1,191 @@
+    #  the edge proxy sets X-Forwarded-For and other proxy headers, which the gateway does not do by default (and
+    #  shouldn't). moreover, in the legacy integration, the `Request` is constructed from the parameters of
+    #  `ProxyListener.forward_request` by the AwsApiListener. However with the gateway, the Request comes directly
+    #  from the server, and `Request.data` may already have been called (consuming the form data), so we will have to
+    #  restore the original request data.
+
+    return ApiInvocationContext(
+        method,
+        path,
+        data,
+        headers,
+    )
+
+
+class ApigatewayRouter:
+    """
+    Simple implementation around a Router to manage dynamic restapi routes (routes added by a user through the
+    apigateway API).
+    """
+
+    router: Router[Handler]
+
+    def __init__(self, router: Router[Handler]):
+        self.router_rules: Dict[str, List[Rule]] = defaultdict(list)
+        self.router = router
+
+    def add_rest_api(self, rest_api: RestApi) -> None:
+        """
+        Adds a route for the given RestApi.
+        :param rest_api: the RestApi to add
+        """
+        # TODO: probably it is better to have a parameterized handler and only one rule, rather than creating a new
+        #  rule for every rest API. the more rules there are, the slower the routing will be. on the other hand,
+        #  regex matching could be just as slow. need to check!
+        api_id = rest_api["id"]
+
+        # add the canonical execute-api handler
+        self.router_rules[api_id].append(
+            self.router.add(
+                "/",
+                host=f"{api_id}.execute-api.<regex('.*'):server>",
+                endpoint=self._restapis_host_handler,
+            )
+        )
+        self.router_rules[api_id].append(
+            self.router.add(
+                "/<path:path>",
+                host=f"{api_id}.execute-api.<regex('.*'):server>",
+                endpoint=self._restapis_host_handler,
+            )
+        )
+
+        # add the localstack-specific _user_request_ handler
+        self.router_rules[api_id].append(
+            self.router.add(
+                f"/restapis/{api_id}/<stage>/_user_request_",
+                endpoint=self._restapis_user_request_handler,
+            )
+        )
+        self.router_rules[api_id].append(
+            self.router.add(
+                f"/restapis/{api_id}/<stage>/_user_request_/<path:path>",
+                endpoint=self._restapis_user_request_handler,
+            )
+        )
+
+    def remove_rest_api(self, rest_api_id: str) -> None:
+        """
+        Removes the given rest api.
+        :param rest_api_id: the rest api to remove
+        """
+        rules = self.router_rules.pop(rest_api_id, [])
+        for rule in rules:
+            self.router.remove_rule(rule)
+
+    def _restapis_handler(self, request: Request, path=None) -> Response:
```

Comment:
nit, `path` is not being used

Reply:
yes good catch! unfortunately it's required by the dispatcher (it resolves URL variables as method parameters)

Reply:
actually strike that, i handed down the url parameters correctly and also used it to hand down the `stage` :-)